### ISSUE TITLE
Unable to generate correct rotations for |e,-α>+|g,α> target state

### ISSUE THREAD

**Issue by Amai-RusseMitsu** [alec-eickbusch/ECD_control#4]
Hi Alec,

I have been using your code to generate pulses and have been aiming to test them on a device. We have started testing simple examples and one such example is the target state |e,-α>+|g,α> or a simple conditional displacement block). Analytically we expect to be able to achieve this target state with a pi/2 pulse and one ECD gate with some beta that achieves the desired alpha. In other words with N=1. 

However upon running this state on the batch optimizer, the result we get for theta is pi and not pi/2. As such the fidelity plateaus at 50%. Changing, use_etas = True seems to result in the fidelity converging to 100% but the rotations that are generated are still not correct ie theta=pi. use_displacement = True improves the result to 77% fidelity with theta \approx 1.8 but this is still not pi/2. 

The attached code is a minimal working example of the problem. 
```
import numpy as np
import qutip as qt 
from ECD_control.ECD_optimization.batch_optimizer import BatchOptimizer
from ECD_control.ECD_optimization.optimization_analysis import OptimizationAnalysis, OptimizationSweepsAnalysis
from ECD_control.ECD_pulse_construction.ECD_pulse_construction import *
import matplotlib.pyplot as plt
from tensorflow.python.client import device_lib

N = 40
alpha = 2.0
psi_t = (qt.tensor(qt.basis(2,1),qt.coherent(N,-alpha)) + qt.tensor(qt.basis(2,0),qt.coherent(N,alpha))).unit()

opt_params = {
'N_blocks' : 1, 
'N_multistart' : 200, 
'epochs' : 200, 
'epoch_size' : 10, 
'learning_rate' : 0.01, 
'term_fid' : 0.999, 
'dfid_stop' : 1e-6, 
'beta_scale' : 4.0, 
'initial_states' : [qt.tensor(qt.basis(2,0),qt.basis(N,0))], 
'target_states' : [psi_t],
'name' : 'cat alpha=%.3f' % alpha,
'filename' : None,
}

opt = BatchOptimizer(**opt_params)
opt.optimize()

#can print info, including the best circuit found.
opt.print_info()
```

May I also ask what use_displacement and use_etas actually does in the batch_optimizer?

Thanks for your help in advance!!

**Comment by alec-eickbusch** [comment]
Hi Russell,

Thank you for pointing out this bug. It turns out there were two issues causing this: 

1. We had set phis[0] = 0 to be fixed since the global phase of the qubit is not important for target states where the qubit and oscillator are disentangled. As your example shows, for target states where the qubit and oscillator are entangled, this phase is important, so I've removed this constraint. phi[0] is now included in the optimization with 41dd36f045ad2e5d54265db5e56e314fc59759f8

2. We had erroneously set the default value of 'use_phase' to be False. This means the optimizer was using only the real part of the state overlap to calculate the fidelity. I've now renamed 'use_phase' to 'real_part_only' to make that parameter less ambiguous and set its default value to False in bf2e8a6586e628f657c01434f5035d684d248988. Now by default the optimizer will use the cost function that takes the absolute value of the overlap squared, not just the real part.

With these two fixes, your example code now works on my machine. I've included it as an example [here](https://github.com/alec-eickbusch/ECD_control/blob/master/examples/entangled_cat.py).

As for use_displacements and use_etas, this was leftover from some experimental gate sets that are different from the version of ECD presented in the paper. I've removed both of these options in 93576efa888e87d1252e2d11582a9850494da1be and cbd44bcc32341e6b8db704acedc506df753f2e36. 

The use_displacements was a gate set which also included a displacement operator in every block. It turns out that this does not shorten circuit depths since these displacements can be commuted to the end of the circuit and only lead to an overall shift of the mean value of \<a\>=\<x\> + i\<p\>. This shift in the mean center-of-mass can instead be implemented in experiment by a single displacement at the end of the circuit. This final displacement was previously included in the code in a complicated way. Now, in a14a912cfc390aefc7b2d30cdd0dae14f64f2d6a, I've added the parameter 'include_final_displacement' which, when True, will also optimize the value of a final displacement operator. An example of where this is important is included in the [entangled cat example](https://github.com/alec-eickbusch/ECD_control/blob/master/examples/entangled_cat.py), where I also show the optimization of a non-centered entangled cat.

use_etas is a gate set that allows the qubit rotations to be around an arbitrary axis on the Bloch sphere (the 'eta' was the angle of the rotation axis relative to the z axis). In experiment, these arbitrary rotations could be achieved using qubits with native sigma_z rotations or by concatenating two rotations around the x-y plane. It turns out this does help to slightly shorten the circuit depth for some quantum control problems, however it may impact the echoing of low-frequency noise. I've moved the eta functionality to a [new branch ](https://github.com/alec-eickbusch/ECD_control/tree/eta_gateset), and if you're interested I'm happy to discuss this more. 

Thanks again, and please feel free to email me or open another github issue if you have any other questions. 

Alec

**Comment by Amai-RusseMitsu** [comment]
Hi Alec,

Thanks for taking the time to explain the variables and the issue, it was very helpful!
